JS: Add helper library for call graph exploration

asgerf opened the pull request:
Exploring the call graph leading to or from a given function should be easier out-of-the-box than it currently is.

This adds a library `semmle.javascript.explore.CallGraph` which can be imported to help define a path query that looks for specific call paths. For example:
```java
/**
 * @kind path-problem
 */

import javascript
import semmle.javascript.explore.CallGraph
import DataFlow

from InvokeNode invoke, FunctionNode target
where callEdge*(invoke, target)
  and isStartOfCallPath(invoke)
  and target.getName() = "myFunction"
select invoke, invoke, target, "Call path to myFunction"
```

There are two other work items I'd like to do, but first I'd like hear what people think of the idea:
- [x] Move `semmle.javascript.dataflow.ForwardExploration` and `semmle.javascript.dataflow.BackwardExploration` into `semmle.javascript.explore`
- [ ] Add a help page about exploratory queries, explaining how to use these libraries.

esbena commented on javascript/ql/src/semmle/javascript/explore/CallGraph.qll:
```diff
@@ -0,0 +1,85 @@
+}
+
+/**
+ * Holds if `invoke` should be used as the starting point of a call path.
```
```suggestion
 * Holds if `fun` should be used as the ending point of a call path.
```

max-schaefer replied:
```suggestion
 * Holds if `invoke` should be used as the end point of a call path.
```

max-schaefer commented on javascript/ql/src/semmle/javascript/explore/CallGraph.qll:
```diff
@@ -0,0 +1,85 @@
+  not callEdge+(rootCall(), invoke)
+}
+
+/** Gets a function contains no calls to other functions. */
```
```suggestion
/** Gets a function that contains no calls to other functions. */
```

max-schaefer commented on javascript/ql/src/semmle/javascript/explore/CallGraph.qll:
```diff
@@ -0,0 +1,85 @@
+ *
+ * There are edges from calls to their callees,
+ * and from functions to their contained calls and in some cases
+ * their inner functions.
```
That last bit is interesting; why do we want those edges?

asgerf replied:
To handle callbacks passed to `forEach` where we don't see the exact call site.

So in a sense, a call edge `A -> B` can be read as `A causes B to be called`.

Even when it's not technically called in the same stack frame (e.g. through `addEventListener`) I still found this to be useful.

max-schaefer replied:
OK, that makes sense. Could you expand the documentation slightly to explain this?

Also, I wonder whether it would make sense to distinguish the two kinds of edges by giving them different labels?

asgerf commented:
Rebased and moved the change note to 1.25. I found myself needing this the other day so I'd like it if we can try and get it merged. Writing the help page turned out to be more time-consuming than expected, so I'd like to merge without that.

asgerf commented:
> Could we remove the aliases to forward/backwards exploration in seemle.javascript.dataflow

It's usually a good idea to have a grace period where both solutions are usable, so I'd like to have both for one release.

asgerf commented:
@esbena you're still requesting changes. Any more requests?
